src/chkreg/chkreg.c

```diff
@@ -0,0 +1,447 @@
+
+    /* Allow only one access at a time. */
+    status2 = sceKernelWaitSema(g_semaId, 1, NULL); // 0x000003C4
+    if (status2 == SCE_ERROR_OK)
+    {
+        if (g_isUMDRegionCodesObtained || (status1 = _sceChkregLookupUMDRegionCodeInfo()) == SCE_ERROR_OK) // 0x000003D8 & 0x000003E8
+        {
+            status1 = _sceChkregCheckRegion(0x80000000, umdMediaType | regionId); // 0x000003F4
+        }
+
+        /* Release acquired sema resource. */
+        status2 = sceKernelSignalSema(g_semaId, 1); // 0x00000404
+        if (status2 != SCE_ERROR_OK)
+        {
+            status1 = SCE_ERROR_SEMAPHORE;
+        }
+    }
+
+    return status1;
+}
+
+// Subroutine sceChkreg_driver_59F8491D - Address 0x00000438
+s32 sceChkregGetPsCode(ScePsCode *pPsCode)
+{
+    s32 status1;
+    s32 status2;
+
+    status1 = SCE_ERROR_SEMAPHORE;
+
+    /* Allow only one access at a time. */
+    status2 = sceKernelWaitSema(g_semaId, 1, NULL);
+    if (status2 == SCE_ERROR_OK)
+    {
+        if ((g_isConsoleIdCertificateObtained || (status1 = _sceChkregLookupConsoleIdCertificate()) == SCE_ERROR_OK)
+            && (status1 = _sceChkregVerifyConsoleIdCertificate()) == SCE_ERROR_OK)
+        {
+            pPsCode->companyCode = g_ConsoleIdCertificate.consoleId.companyCode;
+            pPsCode->productCode = g_ConsoleIdCertificate.consoleId.productCode;
+            pPsCode->productSubCode = g_ConsoleIdCertificate.consoleId.productSubCode;
+            pPsCode->factoryCode = g_ConsoleIdCertificate.consoleId.factoryCode;
+        }
+
+        /* Release acquired sema resource. */
+        status2 = sceKernelSignalSema(g_semaId, 1);
+        if (status2 != SCE_ERROR_OK)
+        {
+            status1 = SCE_ERROR_SEMAPHORE;
+        }
+    }
+
+    return status1;
+}
+
+// Subroutine sceChkreg_driver_9C6E1D34 - Address 0x0000051C
+s32 sceChkreg_driver_9C6E1D34(const u8 *arg0, u8 *arg1)
+{
+    s32 status1;
+    s32 status2;
+
+    status1 = SCE_ERROR_SEMAPHORE;
+
+    /* Allow only one access at a time. */
+    status2 = sceKernelWaitSema(g_semaId, 1, NULL); // 0x00000554
+    if (status2 == SCE_ERROR_OK) // 0x0000055C
+    {
+        u8 *pWorkBuffer = g_pWorkBuffer;
+
+        ((u32 *)pWorkBuffer)[0] = 0x34; // 0x00000570 -- TODO: Length of specific data to hash?
+
+        /* "Prefix" specified input data with 0x14 bytes of Chkreg specific data in the work buffer. */
+
+        // 0x00000574 - 0x0000059C
+        u32 i;
+        for (i = 0; i < CHKREG_UNK9BC_SIZE; i++)
+        {
+            pWorkBuffer[0x4 + i] = g_unk9BC[i];
+        }
+
+        /* Copy 0x20 bytes of input data to work buffer. */
+
+        // 0x000005A0 - 0x000005C0
+        for (i = 0; i < 0x10; i++)
+        {
+            pWorkBuffer[0x4 + CHKREG_UNK9BC_SIZE + i] = arg0[0xD4 + i];
+        }
+
+        // 0x000005C4 - 0x000005E4
+        for (i = 0; i < 0x10; i++)
+        {
+            pWorkBuffer[0x4 + CHKREG_UNK9BC_SIZE + 0x10 + i] = arg0[0x140 + i];
+        }
```

Comment:
How about inline memcpy's?

Reply:
Also, 0xD4 and 0x140 are offsets significant in PSP encrypted binaries, it'd be nice to check what it corresponds to (the mesg_led WIP RE doesn't have the explanations yet but PrxDecrypter might)

Reply:
0xD4 corresponds to [sCheck](https://github.com/uofw/uofw/blob/master/src/loadcore/loadelf.h#L169), 0x140 to [keyData4](https://github.com/uofw/uofw/blob/master/src/loadcore/loadelf.h#L173) in the PspHeader struct. Copy sizes match in the sense that they are not larger than the sizes set in the current structure definition.

(Layout is the same in [pspdecrypt].(https://github.com/John-K/pspdecrypt/blob/master/PrxDecrypter.h#L27))

This could be an indicator that the first argument to this function is actually a `PspHeader *` though with no example call of this API, I'm not sure we should go ahead here and make that change. What do you think?

Reply:
Yeah the comment seems enough.